File: 2770-maximum-number-of-jumps-to-reach-the-last-index/solution.cpp
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

class Solution {
public:
    int maximumJumps(vector<int>& nums, int target) {
        int n = nums.size();
        // dp[i] stores the maximum jumps to reach index i.
        // Initialize with -1 to indicate the index is currently unreachable.
        vector<int> dp(n, -1);
        
        // Base case: 0 jumps to reach the starting index.
        dp[0] = 0;
        
        for (int j = 1; j < n; ++j) {
            for (int i = 0; i < j; ++i) {
                // Check if index i was reachable and if the jump condition is met.
                if (dp[i] != -1 && abs((long long)nums[j] - nums[i]) <= target) {
                    dp[j] = max(dp[j], dp[i] + 1);
                }
            }
        }
        
        // Return the value at the last index.
        return dp[n - 1];
    }
};

File: 2770-maximum-number-of-jumps-to-reach-the-last-index/solution.py
class Solution:
    def maximumJumps(self, nums: list[int], target: int) -> int:
        n = len(nums)
        # dp[i] will store the maximum number of jumps to reach index i.
        # Initialize with -1 to signify that the index is not yet reached.
        dp = [-1] * n
        
        # We start at index 0 with 0 jumps.
        dp[0] = 0
        
        # Iterate through each index from 1 to n-1.
        for j in range(1, n):
            # Check all possible previous indices i from which we can jump to j.
            for i in range(j):
                # If index i is reachable and the jump condition is satisfied.
                if dp[i] != -1 and abs(nums[j] - nums[i]) <= target:
                    # Update dp[j] with the maximum jumps found so far.
                    dp[j] = max(dp[j], dp[i] + 1)
        
        # The answer is the maximum jumps to reach the last index.
        return dp[n - 1]